flow/aggregates.py

```diff
@@ -0,0 +1,509 @@
+
+        :param project:
+            A signac project used to fetch jobs for creating aggregates.
+        :type project:
+            :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+        """
+        if not self._is_aggregate:
+            return _DefaultAggregateStore(project)
+        else:
+            return _AggregatesStore(self, project)
+
+    def __call__(self, func=None):
+        if callable(func):
+            setattr(func, '_flow_aggregate', self)
+            return func
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable, '
+                            f'got {type(func)}.')
+
+
+class _AggregatesStore:
+    """This class holds the information of all the aggregates associated with
+    a :py:class:`aggregator`.
+
+    This is a callable class which, when called, generates all the aggregates.
+    When iterated through it's instance, all the aggregates are yielded.
+
+    :param aggregator:
+        aggregator object associated with this class.
+    :type aggregator:
+        :py:class:`aggregator`
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, aggregator, project):
+        self._aggregator = aggregator
+
+        # We need to create a _MakeAggregate instance associated with this object
+        # in order to generate aggregates whenever necessary.
+        self._make_aggregates = _MakeAggregates.initialize_with_aggregator(aggregator)
+
+        # We need to register the aggregates for this instance using the
+        # project provided.
+        self._aggregates = list()
+        self._aggregate_ids = dict()
+        self._register_aggregates(project)
+
+    def __iter__(self):
+        yield from self._aggregates
+
+    def __getitem__(self, id):
+        "Return an aggregate, if exists, using the id provided"
+        try:
+            return self._aggregate_ids[id]
+        except KeyError:
+            raise LookupError(f'Unable to find the aggregate having id {id} in '
+                              'the FlowProject')
+
+    def __contains__(self, aggregate):
+        """Return whether an aggregate is stored in the this
+        instance of :py:class:`_AggregateStore`
+
+        :param aggregate:
+            An aggregate of jobs.
+        :type aggregate:
+            tuple of :py:class:`signac.contrib.job.Job`
+        """
+        return get_aggregate_id(aggregate) in self._aggregate_ids
+
+    def __len__(self):
+        return len(self._aggregates)
+
+    def __eq__(self, other):
+        return type(self) == type(other) and \
+               self._aggregate_ids.keys() == other._aggregate_ids.keys() and \
+               self._aggregator == other._aggregator
```

Comment:
The last part is unnecessary. If all the aggregate ids are the same then the store is the same for all intents and purposes, unless you think that multiple forms of aggregation may lead to the same exact aggregates.

Reply:
I believe we should be extra sure, there may be a case when we only have a single job in the project and two operations which groups them in groups of 2, groups of 3. Then we won't be differentiating between them. This is completely theoretical and I'll do as you suggest